File: clojure/src/aoc24/day03.clj
(ns aoc24.day03
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn parse-input [file]
  (str/trim (slurp (io/resource file))))

(defn execute-mul [input enabled?]
  (let [matches (re-seq #"mul\((\d+),(\d+)\)" input)]
    (reduce (fn [sum [_ a b]]
              (if enabled?
                (+ sum (* (parse-long a) (parse-long b)))
                sum))
            0
            matches)))

(def input (parse-input "aoc24/day03.txt"))

(defn part-1
  "Run with bb -x aoc24.day03/part-1"
  [_]
  (prn (execute-mul input true)))

(defn- find-next-instruction [input]
  (let [do-index (str/index-of input "do()")
        dont-index (str/index-of input "don't()")]
    (cond
      (and (nil? do-index) (nil? dont-index)) nil
      (nil? do-index) [:dont dont-index]
      (nil? dont-index) [:do do-index]
      (< do-index dont-index) [:do do-index]
      :else [:dont dont-index])))

(defn- process-chunk [input enabled? sum]
  (if-let [[instruction-type index] (find-next-instruction input)]
    (let [chunk (subs input 0 index)
          remaining-input (subs input (+ index (if (= instruction-type :do) 4 6)))]
      [(+ sum (execute-mul chunk enabled?))
       (= instruction-type :do)
       remaining-input])
    [(+ sum (execute-mul input enabled?)) enabled? nil]))

(defn part-2
  "Run with bb -x aoc24.day03/part-2"
  [_]
  (loop [input input
         sum 0
         enabled? true]
    (let [[new-sum new-enabled? remaining-input] (process-chunk input enabled? sum)]
      (if (nil? remaining-input)
        (prn new-sum)
        (recur remaining-input new-sum new-enabled?)))))

(comment

  (part-1 nil)
  (part-2 nil)

  'comment)  
  

File: java/src/aoc24/Day03.java
package aoc24;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day03 {

    public static void main(String[] args) throws IOException {
        Path inputPath = Path.of("resources", "aoc24", "day03.txt");
        String input = Files.readString(inputPath).trim();

        long part1Result = part1(input);
        System.out.println("Part 1: " + part1Result);

        long part2Result = part2(input);
        System.out.println("Part 2: " + part2Result);
    }

    public static long executeMul(String input, boolean enabled) {
        Pattern pattern = Pattern.compile("mul\\((\\d+),(\\d+)\\)");
        Matcher matcher = pattern.matcher(input);
        long sum = 0;
        while (matcher.find()) {
            if (enabled) {
                long a = Long.parseLong(matcher.group(1));
                long b = Long.parseLong(matcher.group(2));
                sum += a * b;
            }
        }
        return sum;
    }

    public static long part1(String input) {
        return executeMul(input, true);
    }

    public static long part2(String input) {
        long sum = 0;
        boolean enabled = true;
        int doIndex = input.indexOf("do()");
        int dontIndex = input.indexOf("don't()");

        while (doIndex != -1 || dontIndex != -1) {
            int nextIndex = -1;
            if (doIndex != -1 && (dontIndex == -1 || doIndex < dontIndex)) {
                nextIndex = doIndex;
            } else if (dontIndex != -1) {
                nextIndex = dontIndex;
            }

            sum += executeMul(input.substring(0, nextIndex), enabled);
            enabled = (nextIndex == doIndex);
            input = input.substring(nextIndex + (nextIndex == doIndex ? 4 : 6));

            doIndex = input.indexOf("do()");
            dontIndex = input.indexOf("don't()");
        }

        sum += executeMul(input, enabled);
        return sum;
    }
}
